### PR TITLE
Consolidate Garbage Collection docs

### LINKED ISSUE #4322: Better docs for garbage collection
**This is a...** 
- [X] Feature Request
- [ ] Bug Report

## Problem 

K8s docs discuss garbage collection in different articles:

- https://kubernetes.io/docs/concepts/cluster-administration/kubelet-garbage-collection/
- https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/
- https://kubernetes.io/docs/admin/kube-controller-manager/
- https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/

Multiple topics show little to no awareness of each other. 

Descriptions of garbage collection sometimes conflict. 

No single article fully describes garbage collection as a concept or comprehensively demonstrates its implementation. 

There is currently no task for garbage collection: only two concept articles and some incidental reference material.

## Proposed Solution

1. Consolidate conceptual content at https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/.

  See _Additional information_ for more details about conceptual content.

  Explain garbage collection in Kubernetes as a whole: what it is; why and how different components (kubelet, kube-controller-manager) handle it separately; and interactions and dependencies between components' garbage collection functions.

  Provide an end-to-end description of garbage collection in a typical production environment.

2. Create a new **task article**.

  Explain how to set up garbage collection in a typical production environment: 
    - Which components to configure, and in what order, for default function
    - How to assess the health of garbage collection for a pod

## Pages to update

- https://kubernetes.io/docs/concepts/cluster-administration/kubelet-garbage-collection/
- https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/
- https://kubernetes.io/docs/admin/kube-controller-manager/

## Additional information

Different aspects of garbage collection appear to be handled by kubelet and kube-controller-manager. It's unclear from the documentation whether these components' garbage collection functions interact and the extent to which their functions are interdependent.

It's unclear which garbage control functions are available in which versions of Kubernetes:
  -  https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/:
  "Note: Garbage collection is a beta feature and is enabled by default in Kubernetes version 1.4 and later." 
  - Is garbage collection also a beta feature in kubelet and kube-controller-manager?

Different topics' descriptions of garbage collection don't match:
  - https://kubernetes.io/docs/concepts/cluster-administration/kubelet-garbage-collection/ describes garbage collection as a function of kubelet.
  - https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/ describes a controller as "the Kubernetes garbage collector"
  - https://kubernetes.io/docs/admin/kube-controller-manager/ also sets some attributes for garbage collection.

Some of the most helpful material for understanding garbage collection mentions it only incidentally: 
  - https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/

Multiple topics break the style guide by making statements about the future:

   - https://kubernetes.io/docs/concepts/cluster-administration/kubelet-garbage-collection/: "Some kubelet Garbage Collection features in this doc will be replaced by kubelet eviction in the future."
  - https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/: "Kubernetes version 1.7 will add an admission controller that controls user access to set blockOwnerDeletion to true based on delete permissions on the owner object, so that unauthorized dependents cannot delay deletion of an owner object."

### PR DESCRIPTION
**Pull request by shannonxtreme**
Fixes #4322 

## Overview

Consolidate and improve docs on Garbage Collection. 

The following tasks should be completed before review: 

* [x] Create a GC concept page
* [x] Create a Finalizers concept page
* [x] Create GC glossary term
* [x] Create finalizers glossary term
* [ ] Create task for configuring kubelet GC
* [ ] Create task for configuring cascading deletion
* [ ] Create task for using Finalizers to control deletion (or link to blog post)

/sig docs
/assign
/cc @sftim 
/do-not-merge work-in-progress

### CODE REVIEW

**Comment by shannonxtreme** [comment]
Tim: I haven't added glossary links and all that fun stuff yet. I'm working on drafting the tasks based on existing content so that all our hyperlinks have actual destinations. ETA end of the week on the remaining tasks :)

**Comment by sftim** [comment]
/wg api-expression
for parts of this

File: content/en/docs/concepts/workloads/controllers/garbage-collection.md [hunk 1]
```diff
@@ -5,180 +5,137 @@ weight: 60
...
+  * [Unused containers and container images](#containers-images)
+  * [Failed pods](/docs/concepts/workloads/pods/pod-lifecycle/#pod-garbage-collection)
+  * [Dynamically provisioned PersistentVolumes with a StorageClass reclaim policy of Delete](/docs/concepts/storage/persistent-volumes/#delete)
+  * [Stale or expired CertificateSigningRequests (CSRs)](/reference/access-authn-authz/certificate-signing-requests/#request-signing-process)
+  * Nodes deleted in the following scenarios:
+    * On a cloud when the cluster uses a [cloud controller manager](/docs/concepts/architecture/cloud-controller/)
+    * On-premises when the cluster uses an addon similar to a cloud controller
+      manager
+  * [Node Lease objects](/docs/concepts/architecture/nodes/#heartbeats)
+  * [Completed Jobs](/docs/concepts/workloads/controllers/ttlafterfinished/)
```
Comment (sftim) [inline content/en/docs/concepts/workloads/controllers/garbage-collection.md:21]: As this is in currently part of workloads section, _this_ page should either:
- focus on cleanup for completed jobs
- move somewhere else

File: content/en/docs/reference/glossary/finalizers.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+- fundamental
+- operation
+---
+Resource keys that specify what actions to take before deleting specific resources.
+
+<!--more-->
+
+Finalizers exist as lists in the `metadata.finalizers` field of a resource
+definition and specify what a {{<glossary_tooltip text="controller" term_id="controller">}}
+must do before it can delete the resource.
```
Comment (sftim) [inline content/en/docs/reference/glossary/finalizers.md:20]: The glossary entry should be singular: “finalizer”.

File: content/en/docs/reference/glossary/garbage-collection.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+- operation
+---
+ A collective term for the various mechanisms Kubernetes
+ uses to clean up cluster resources.
+
+<!--more-->
+
+Kubernetes uses garbage collection to clean up resources like unused containers
+and images, failed pods, deleted nodes, orphaned objects, expired resources, and
+completed Jobs.
```
Comment (sftim) [inline content/en/docs/reference/glossary/garbage-collection.md:22]: These examples could be links to pages (tasks or concepts) about each one.

**Comment by sftim** [comment]
/sig apps
for Job cleanup

**Comment by shannonxtreme** [comment]
@sftim: for configuring container and image garbage collection, we already have a task for [configuring kubelet using a file](https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/) or [in a live node](https://kubernetes.io/docs/tasks/administer-cluster/reconfigure-kubelet/). Maybe instead of a new task that duplicates some of the content, we can link to that page and to the [kubeletconfiguration reference](https://kubernetes.io/docs/tasks/administer-cluster/reconfigure-kubelet/) so users can build off that?

File: content/en/docs/concepts/workloads/controllers/garbage-collection.md (hunk 1 above)
Comment (shannonxtreme) [reply content/en/docs/concepts/workloads/controllers/garbage-collection.md:21]: By this page, do you mean the TTLafterfinished page?

**Pull request closed without merging**